crates/valence_server/src/placement.rs

```diff
@@ -0,0 +1,939 @@
+                    | BlockKind::PackedIce
+                    | BlockKind::FrostedIce => true,
+                    _ => false,
+                },
+            }
+        }
+        _ => unimplemented!(),
+    }
+}
+
+pub fn can_be_placed_on_side(this: BlockState, of: BlockState) -> bool {
+    match PlacementRule::from_kind(of.to_kind()) {
+        PlacementRule::Basic
+        | PlacementRule::FrostedIce
+        | PlacementRule::Doors
+        | PlacementRule::Composter
+        | PlacementRule::Observer => return true,
+        PlacementRule::Trapdoors => {
+            if of.get(PropName::Half).unwrap() == PropValue::Top
+                && of.get(PropName::Open).unwrap() == PropValue::False
+            {
+                return true;
+            }
+        }
+        PlacementRule::Scaffolding
+        | PlacementRule::EndRod
+        | PlacementRule::MobHeads
+        | PlacementRule::ChorusPlant => {
+            return false;
+        }
+        PlacementRule::Snow => match of.get(PropName::Layers).unwrap() {
+            PropValue::_2
+            | PropValue::_3
+            | PropValue::_4
+            | PropValue::_5
+            | PropValue::_6
+            | PropValue::_7 => return false,
+            _ => {}
+        },
+        _ => {}
+    }
+
+    match of.to_kind() {
+        BlockKind::Azalea | BlockKind::FloweringAzalea => {
+            return false;
+        }
+        _ => {}
+    }
+
+    match PlacementRule::from_kind(this.to_kind()) {
+        PlacementRule::Buttons | PlacementRule::Lever => {
+            match PlacementRule::from_kind(of.to_kind()) {
+                PlacementRule::ChorusFlower => true,
+                PlacementRule::Snow => of.get(PropName::Layers).unwrap() == PropValue::_8,
+                _ => false,
+            }
+        }
+        _ => unimplemented!(),
+    }
+}
+
+pub fn can_be_placed_on_bottom(this: BlockState, of: BlockState) -> bool {
+    todo!()
+}
+
+//TODO: Builder
+#[derive(Clone)]
+pub struct PlaceBlockCommand {
+    pub block_kind: BlockKind,
+    pub interact_block_event: InteractBlockEvent,
+    pub look: Look,
+    pub ignore_collisions: bool,
+}
+
+impl Command for PlaceBlockCommand {
+    fn apply(self, world: &mut bevy_ecs::world::World) {
+        let Self {
+            block_kind,
+            interact_block_event:
+                InteractBlockEvent {
+                    client: _,
+                    hand: _,
+                    position: interact_pos,
+                    face,
+                    cursor_pos,
+                    head_inside_block,
+                    sequence: _,
+                },
+            look,
+            ignore_collisions,
+        } = self;
+
+        let mut system_state =
+            SystemState::<(Query<&mut ChunkLayer>, Query<&Hitbox, With<LivingEntity>>)>::new(world);
+        let (mut layers, entity_hitboxes) = system_state.get_mut(world);
+
+        let mut layer = layers.single_mut();
```

Comment:
You can't do that, you either need to get the layer the client is on or ask for a layer to place the block

Reply:
I'm not sure what you mean, this is basically just copied from the `building` example.

Reply:
You can't call `layers.single_mut()` because it'll panic if there are more than one entity with a `ChunkLayer`. It's there in the example code for simplicity.

Reply:
ah yea I had thought for a moment, but though the example will have a reason.. and yea it has a reason.
Thanks for pointing it out :D